src/erlfmt_ast.erl

```diff
@@ -0,0 +1,1086 @@
+                    {macro_call, Pos, Name, Args}
+            end;
+        type_application ->
+            [[Name], Args] = st_subtrees_to_erlfmt(Node),
+            case Name of
+                {remote, NPos, M, N} ->
+                    exit(remote_type),
+                    {remote, NPos, [M, N, Args]};
+                _ ->
+                    exit({local_type, Name}),
+                    {call, Pos, Name, Args}
+            end;
+        user_type_application ->
+            exit({user_type, Node}),
+            [[Name], Args] = st_subtrees_to_erlfmt(Node),
+            {call, Pos, Name, Args};
+        annotated_type ->
+            [[Name], [Type]] = st_subtrees_to_erlfmt(Node),
+            {op, Pos, '::', Name, Type};
+        typed_record_field ->
+            [[Body], [Type]] = st_subtrees_to_erlfmt(Node),
+            {op, Pos, '::', Body, Type};
+        %% for expressions with clauses we cannot first revert the clauses
+        %% separately, because the revert of the expression looks at the
+        %% clauses and crashes if they are on erlfmt format
+        if_expr ->
+            {'if', Pos, [
+                st_clause_to_erlfmt(C, 'if')
+                || C <- erl_syntax:if_expr_clauses(Node)
+            ]};
+        case_expr ->
+            {'case', Pos, st_to_erlfmt(erl_syntax:case_expr_argument(Node)), [
+                st_clause_to_erlfmt(C, 'case')
+                || C <- erl_syntax:case_expr_clauses(Node)
+            ]};
+        receive_expr ->
+            Clauses = [
+                st_clause_to_erlfmt(C, 'case')
+                || C <- erl_syntax:receive_expr_clauses(Node)
+            ],
+            case erl_syntax:receive_expr_timeout(Node) of
+                none ->
+                    {'receive', Pos, Clauses};
+                Timeout ->
+                    TBody = [
+                        st_to_erlfmt(E)
+                        || E <- erl_syntax:receive_expr_action(Node)
+                    ],
+                    {'receive', Pos, Clauses, st_to_erlfmt(Timeout), TBody}
+            end;
+        try_expr ->
+            [Body0] = erl_syntax:try_expr_body(Node),
+            Body = st_to_erlfmt(Body0),
+            Clauses = [
+                st_clause_to_erlfmt(C, 'case')
+                || C <- erl_syntax:try_expr_clauses(Node)
+            ],
+            Handlers = [
+                st_clause_to_erlfmt(C, 'try')
+                || C <- erl_syntax:try_expr_handlers(Node)
+            ],
+            After = [
+                st_to_erlfmt(E)
+                || E <- erl_syntax:try_expr_after(Node)
+            ],
+            CAnno = dummy_anno(),
+            {'try', Pos, Body, {clauses, CAnno, Clauses}, {clauses, CAnno, Handlers}, After};
+        comment ->
+            {comment, Pos, erl_syntax:comment_text(Node)};
+        _ ->
+            st_to_erlfmt_1(Node)
+    end.
+
+st_to_erlfmt_1(Node) ->
+    %% TODO: should we convert full erl_syntax pos+annotation to erlfmt anno?
```

Comment:
It seems like you are doing Anno conversions here, so maybe this TODO is already done or I don't understand it, sorry?

Reply:
erl_syntax nodes have a separate annotation field (because the old annotation field used to be only for line numbers), and that could contain arbitrary "user data", so the question is whether to pick up any such annotations and inject them into the erlfmt annotation when going from st to erlfmt, or just abandon them (as done now).

Reply:
Aha, well I can say that only the relative line numbers are taken into account by the formatting algorithm.
So then the question would be, what would be the use case to preserve them and is that worth the effort.